src/librustc_mir/transform/preallocate_gc_contents.rs

```diff
@@ -0,0 +1,77 @@
+#![allow(dead_code)]
+#![allow(unused_imports)]
+use crate::transform::{MirPass, MirSource};
+use crate::util::patch::MirPatch;
+use rustc_infer::infer::TyCtxtInferExt;
+use rustc_middle::mir::*;
+use rustc_middle::traits;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_span::DUMMY_SP;
+use rustc_trait_selection::traits::predicate_for_trait_def;
+use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt;
+
+static VEC_CONSTRUCTOR: &str = "std::vec::Vec::<T>::new";
+
+pub struct GcPreallocator;
+
+impl<'tcx> MirPass<'tcx> for GcPreallocator {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, source: MirSource<'tcx>, body: &mut BodyAndCache<'tcx>) {
+        let param_env = tcx.param_env(source.def_id()).with_reveal_all();
+        let mut patch = MirPatch::new(body);
+
+        for (bb, bb_data) in body.basic_blocks().iter_enumerated() {
+            let terminator = bb_data.terminator();
+            if let TerminatorKind::Call {
+                func: ref op,
+                ref args,
+                ref destination,
+                cleanup,
+                from_hir_call,
+            } = terminator.kind
+            {
+                if let ty::FnDef(callee_def_id, substs) = op.ty(&**body, tcx).kind {
+                    let name = tcx.def_path_str(callee_def_id);
+                    if name == VEC_CONSTRUCTOR {
```

Comment:
You use a string comparison because you didn't want to make all kinds of constructors in the std lib lang items, right?

Reply:
Are we sure this captures all vector constructors?

In other words, are we certain `std::vec::Vec::<T>::new` is canonical at this stage in the MIR? I'd hope so, but I'm unsure.

Reply:
> You use a string comparison because you didn't want to make all kinds of constructors in the std lib lang items, right?

I'd be lying if I said this was the primary reason. However it seems that you can't attach lang items to methods in impl blocks (`tcx.lang_items().my_lang_item().unwrap()` will fail!).

Reply:
> Are we sure this captures all vector constructors?

No, and I'm not particularly satisfied with it either. It's very much a bodge. When I turn this into a merge-able PR, I think a good approach to tidying this up is to make the `impl Vec` block a lang item and then search its methods in the HIR by string, this gets around any problems with non-canonicalised names (or even `as` aliasing).

Reply:
> When I turn this into a merge-able PR, I think a good approach to tidying this up is to make the impl Vec block a lang item and then search its methods in the HIR by string

Let's do that in a later PR -- we shouldn't expect to make every PR perfect, and I'd like to get some measurements, and this PR merged, sooner rather than later.